auraed/src/init/system_runtimes/mod.rs

```diff
@@ -28,18 +34,82 @@
  *                                                                            *
 \* -------------------------------------------------------------------------- */
 use super::InitError;
+use anyhow::{anyhow, Context};
 pub(crate) use cell_system_runtime::CellSystemRuntime;
 pub(crate) use container_system_runtime::ContainerSystemRuntime;
 pub(crate) use daemon_system_runtime::DaemonSystemRuntime;
 pub(crate) use pid1_system_runtime::Pid1SystemRuntime;
+use tokio::net::{TcpListener, UnixListener};
+use tokio_stream::wrappers::{TcpListenerStream, UnixListenerStream};
 use tonic::async_trait;
+use tracing::{info, trace};
 
 mod cell_system_runtime;
 mod container_system_runtime;
 mod daemon_system_runtime;
 mod pid1_system_runtime;
 
+/// A [SocketStream] can represent either a TCP or Unix socket stream.
+#[derive(Debug)]
+pub enum SocketStream {
+    /// Contains a stream for listening over a TCP socket.
+    Tcp {
+        /// A stream that listens over a TCP socket.
+        stream: TcpListenerStream,
+    },
+
+    /// Contains a stream for listening over a Unix socket.
+    Unix {
+        /// A stream that listens over a Unix socket.
+        stream: UnixListenerStream,
+    },
+}
+
 #[async_trait]
 pub(crate) trait SystemRuntime {
-    async fn init(self, verbose: bool) -> Result<(), InitError>;
+    async fn init(
+        self,
+        verbose: bool,
+        socket_address: Option<String>,
+    ) -> Result<SocketStream, InitError>;
+}
+
+async fn create_unix_socket_stream(
+    socket_path: PathBuf,
+) -> Result<SocketStream, InitError> {
+    let _ = std::fs::remove_file(&socket_path);
+    let sock_path = Path::new(&socket_path)
+        .parent()
+        .ok_or_else(|| anyhow!("unable to find socket path"))?;
```

Comment:
The wording on this error made me think that the directory needs to exist already. Maybe something like "not a valid socket path" would be better?

Reply:
ack